scripts/generate_rust_target.rs

```diff
@@ -0,0 +1,317 @@
+
+        let stdin = std::io::stdin();
+        let mut handle = stdin.lock();
+        let mut line = String::new();
+
+        loop {
+            line.clear();
+
+            if handle.read_line(&mut line).unwrap() == 0 {
+                break;
+            }
+
+            if line.starts_with('#') {
+                continue;
+            }
+
+            let (key, value) = line.split_once('=').expect("Missing `=` in line.");
+            result.insert(key.to_string(), value.trim_end_matches('\n').to_string());
+        }
+
+        KernelConfig(result)
+    }
+
+    /// Does the option exist in the configuration (any value)?
+    ///
+    /// The argument must be passed without the `CONFIG_` prefix.
+    /// This avoids repetition and it also avoids `fixdep` making us
+    /// depending on it.
+    fn has(&self, option: &str) -> bool {
+        let option = "CONFIG_".to_owned() + option;
+        self.0.contains_key(&option)
+    }
+}
+
+fn main() {
+    let cfg = KernelConfig::from_stdin();
+    let mut ts = TargetSpec::new();
+
+    let pre_link_args = vec![(
+        "gcc".to_string(),
+        Value::Array(vec![
+            Value::String("-Wl,--as-needed".to_string()),
+            Value::String("-Wl,-z,noexecstack".to_string()),
+        ]),
+    )];
+
+    let pre_link_args_32 = vec![(
+        "gcc".to_string(),
+        Value::Array(vec![
+            Value::String("-Wl,--as-needed".to_string()),
+            Value::String("-Wl,-z,noexecstack".to_string()),
+            Value::String("-m32".to_string()),
+        ]),
+    )];
+
+    let pre_link_args_64 = vec![(
+        "gcc".to_string(),
+        Value::Array(vec![
+            Value::String("-Wl,--as-needed".to_string()),
+            Value::String("-Wl,-z,noexecstack".to_string()),
+            Value::String("-m64".to_string()),
+        ]),
+    )];
+
+    let stack_probes = vec![("kind".to_string(), Value::String("none".to_string()))];
+
+    if cfg.has("ARM") {
+        ts.push("arch", "arm");
+        ts.push(
+            "data-layout",
+            "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64",
+        );
+        ts.push("dynamic-linking", true);
+        ts.push("crt-static-respected", true);
+        ts.push("executables", true);
+        ts.push("features", "+strict-align,+v6");
+        ts.push("has-elf-tls", true);
+        ts.push("has-rpath", true);
+        ts.push("llvm-target", "arm-unknown-linux-gnueabi");
+        ts.push("max-atomic-width", 64);
+        ts.push("pre-link-args", pre_link_args);
+        ts.push("target-family", "unix");
+        ts.push("target-mcount", "\\u0001__gnu_mcount_nc");
+        ts.push("target-pointer-width", "32");
+    } else if cfg.has("ARM64") {
+        ts.push("arch", "aarch64");
+        ts.push(
+            "data-layout",
+            "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128",
+        );
+        ts.push("disable-redzone", true);
+        ts.push("emit-debug-gdb-scripts", false);
```

Comment:
This one should probably be moved out of the target match.

Reply:
Yeah, but please see the commit message, i.e. the intention of this PR is to keep the targets exactly the same as they are now, so that this commit is focused only about introducing the script, and then clean things up. Same for the other bits.

Reply:
Makes sense.